lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -630,17 +630,26 @@ Type TypeChecker::typeCheckParameterDefault(Expr *&defaultValue,
   // Check whether generic parameters are only mentioned once in
   // the anchor's signature.
   {
-    auto anchorTy = anchor->getInterfaceType()->castTo<GenericFunctionType>();
+    auto anchorTy = anchor->getInterfaceType()->castTo<AnyFunctionType>();
 
-    // Reject if generic parameters are used in multiple different positions
-    // in the parameter list.
+    if (anchor->hasCurriedSelf())
+      anchorTy = anchorTy->getResult()->castTo<AnyFunctionType>();
+
+    SmallPtrSet<CanType, 4> inferrableParams;
+    collectReferencedGenericParams(paramInterfaceTy, inferrableParams);
 
     llvm::SmallVector<unsigned, 2> affectedParams;
     for (unsigned i : indices(anchorTy->getParams())) {
-      const auto &param = anchorTy->getParams()[i];
 
-      if (containsTypes(param.getPlainType()))
-        affectedParams.push_back(i);
+      SmallPtrSet<CanType, 4> referencedGenericParams;
+      collectReferencedGenericParams(anchorTy->getParams()[i].getPlainType(), referencedGenericParams);
+
+      for (auto can : referencedGenericParams) {
+        if (inferrableParams.contains(can)) {
```

Comment:
Note that `referencedGenericParams` can also contain DependentMemberTypes, I think we ought to only check for `GenericTypeParamType`s here to allow e.g:

```swift
protocol P {
  associatedtype X
}
extension Int: P {
  typealias X = Int
}

struct S {
  func foo<T: P>(_ x: (T, T.X) = (0, 0), _ y: T.X) {}
}
```

Reply:
I appreciate the detailed insights. Sorry for missing all these edge cases 😕